sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -658,6 +658,41 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
     }
   }
 
+  private[this] def stringToDecimal(str: UTF8String, decimalType: DecimalType): Decimal = {
+    val bigDecimal = try {
+      // According the benchmark test,  `s.toString.trim` is much faster than `s.trim.toString`.
+      // Please refer to https://github.com/apache/spark/pull/26640
+      new JavaBigDecimal(str.toString.trim)
+    } catch {
+      case _: NumberFormatException =>
+        if (ansiEnabled) {
+          throw new NumberFormatException(s"invalid input syntax for type numeric: $str")
+        } else {
+          null
+        }
+    }
+
+    if (bigDecimal != null) {
+      val precision = if (bigDecimal.scale < 0) {
+        bigDecimal.precision - bigDecimal.scale
+      } else {
+        bigDecimal.precision
+      }
+
+      if (precision > DecimalType.MAX_PRECISION) {
```

Comment:
Please leave some comments here about why we need this redundant check?

Reply:
+1, let's explain which part is slow.